### PR TITLE
Do not call into MsQuic inside a lock

### LINKED ISSUE #59345: [QUIC] Do not call msquic in a lock
Check all places where we call msquic and make sure we're not doing that in a lock since it can cause a deadlock.
For example: 
https://github.com/dotnet/runtime/blob/4d411739ce1d9db68dfe12227663479c9d71c50d/src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs#L491
https://github.com/dotnet/runtime/blob/4d411739ce1d9db68dfe12227663479c9d71c50d/src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs#L486

Add asserts before msquic calls that the lock is not taken, e.g.:
https://github.com/dotnet/runtime/blob/4d411739ce1d9db68dfe12227663479c9d71c50d/src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs#L848

### PR DESCRIPTION
**Pull request by rzikm**
Also added `Debug.Assert` calls to ensure that any calls to MsQuic APIs inside a lock are caught in time.

Fixes #59345

### CODE REVIEW

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 1]
```diff
@@ -845,6 +863,7 @@ private void Dispose(bool disposing)
 
         private void EnableReceive()
         {
+            Debug.Assert(!Monitor.IsEntered(_state), "!Monitor.IsEntered(_state)");
```
Comment (danmoseley) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:866]: Instead of repeating all this, it might be a little more readable to just call a method?

```c#
[Conditional("DEBUG")]
internal void AssertMonitorNotEntered(object obj)
{
     Debug.Assert(!Monitor.IsEntered(obj), "Monitor was unexpectedly held");
}
```

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 2]
```diff
@@ -482,22 +487,32 @@ internal override ValueTask<int> ReadAsync(Memory<byte> destination, Cancellatio
...
+
+            // methods below need to be called outside of the lock
+            if (bytesRead > -1)
+            {
+                ReceiveComplete(bytesRead);
 
-                    return new ValueTask<int>(taken);
+                if (reenableReceive)
+                {
+                    EnableReceive();
```
Comment (stephentoub) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:512]: If we're calling this outside of the lock, is there something else that ensures we're not racing with another thread to disable receives?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:866]: (If we do https://github.com/dotnet/runtime/issues/65965, it'll also be redundant in this case, as with or without the message you'd get the same assert.)

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs (hunk 2 above)
Comment (CarnaViire) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:512]: TL;DR: RECV event is the only thing disabling receives. It seems to me that it should be ok due to additional guard from msquic side (no new RECV event will come until ReceiveComplete+EnableReceive is called). But I don't like how fragile it looks 😢

Note that the change is only in a branch for "IndividualReadComplete" state. It means it happens AFTER some data already arrived in RECV event.

We do have a guard for parallel reads which looks exactly at ReadState. So it will throw if ReadAsync is called while state is PendingRead (no data available and there's already a waiting read).

By moving ReceiveComplete+EnableReceive out of the lock, we allow a time where state is already changed from "IndividualReadComplete" to "None", but ongoing "first" ReceiveAsync function is not exited yet and ReceiveComplete+EnableReceive are not called yet (but all data is already copied). So a second ReceiveAsync might enter. It would see state "None" which would mean data is not available. If it grabs the lock before ReceiveComplete+EnableReceive are called (i.e. RECV event is still not possible), it would change the state to PendingRead (waiting for data), store the destination buffer reference, register cancellation and return a task to wait. All the things "None" branch touches are not touched in the remainder of the first ReceiveAsync (between exiting from the lock and returning new value task). 

The bottom line is I don't think there's any problem in ReadAsync vs HandleEventRecv race. But we might want to rethink guard against parallel reads.

Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:512]: So if I understand correctly, the only think that can go wrong with this change is parallel ReadAsync operations, which user code is not supposed to do anyway...

Should I add the guards against parallel operations in this PR, or should we do that separately? @ManickaP @stephentoub , thoughts?

Comment (CarnaViire) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:512]: I would say leave it as is for now. We have a separate issue for proper guards against parallel reads and writes https://github.com/dotnet/runtime/issues/52627

**Comment by ManickaP** [comment]
There's a segmentation fault from the H/3 stress run, in artifacts is dump. Could you please investigate the cause? If it's not related to this change, we can merge, but we need to file an issue for it at least.

**Comment by ManickaP** [comment]
@rzikm https://github.com/dotnet/runtime/pull/67230 seems to be getting closer to stable, so if you're not successful with local run, you might opt to wait for this to get in main.

**Comment by rzikm** [comment]
The [second run of the http-stress](https://dev.azure.com/dnceng/public/_build/results?buildId=1688169&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=1451f5f3-0108-5a08-5b92-e984b2a85bbd) did not crash, are we good to move forward, @ManickaP?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs [hunk 3]
```diff
@@ -149,7 +149,7 @@ public MsQuicConnection(IPEndPoint localEndPoint, IPEndPoint remoteEndPoint, MsQ
 
             try
             {
-                Debug.Assert(!Monitor.IsEntered(_state));
+                Debug.Assert(!Monitor.IsEntered(_state), "!Monitor.IsEntered(_state)");
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs:152]: What led you to even expand more the messages in asserts after Stephen's comment?
It'll become redundant.

Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs:152]: I thought it would be more useful to have a message handy when something crashes, I thought we generally tend to include the message with the assert in this repo

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 4]
```diff
@@ -482,22 +487,32 @@ internal override ValueTask<int> ReadAsync(Memory<byte> destination, Cancellatio
...
                     else if (_state.ReceiveIsFinal)
                     {
                         // This was a final message and we've consumed everything. We can complete the state without waiting for PEER_SEND_SHUTDOWN
                         _state.ReadState = ReadState.ReadsCompleted;
                     }
+                }
+            }
+
+            // methods below need to be called outside of the lock
+            if (bytesRead > -1)
```
[the remark this answers is not in the document]
Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:506]: it will be -1 if `initialReadState != ReadState.IndividualReadComplete`, I added it in order for `EnableReceive` and `ReceiveComplete` to be called (outside of the lock) only if they would've been called in the previous version (inside of the lock)

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs (hunk 3 above)
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs:152]: It's exactly the other way around, just grep for `Debug.Assert` in the repo. And if it contains the message, I mostly see some additional info and not just the copy of the condition.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:506]: I see now, it's its initial value set before the lock.

**Review by ManickaP — APPROVED** [review]
Just nits, otherwise good to merge, thanks!

**Comment by nibanks** [comment]
I know this is already merged, but you're totally allowed to call **async** MsQuic functions under lock. For instance, most (all?) datapath functions would be fine. Every function annotated with `_IRQL_requires_max_(DISPATCH_LEVEL)` _should_ be Ok to call, because it's already designed to be called at DISPATCH level in kernel, which means it cannot be blocked by anything else inside MsQuic that would cause a deadlock.
